Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs

```diff
@@ -1,46 +1,276 @@
+            _activeAnalyzers.Remove(uid);
+        }
+
+        /// <summary>
+        /// Fetches new data when a user clicks on the refresh button
+        /// </summary>
+        private void OnRefreshMessage(EntityUid uid, GasAnalyzerComponent component, GasAnalyzerRefreshMessage message)
+        {
+            if (message.Session.AttachedEntity is not {Valid: true})
+                return;
+
+            UpdateAnalyzer(uid, component);
+        }
+
+        /// <summary>
+        /// Close the UI, turn the analyzer off, and don't update when it's dropped
+        /// </summary>
+        private void OnDropped(EntityUid uid, GasAnalyzerComponent component, DroppedEvent args)
+        {
+            DisableAnalyzer(uid, component, args.User);
+        }
+
+        /// <summary>
+        /// Closes the UI, sets the icon to off, and removes it from the update list
+        /// </summary>
+        private void DisableAnalyzer(EntityUid uid, GasAnalyzerComponent? component = null, EntityUid? user = null)
+        {
+            if (!Resolve(uid, ref component))
+                return;
+
+            if (user != null && TryComp<ActorComponent>(user, out var actor))
+                _userInterface.TryClose(uid, GasAnalyzerUiKey.Key, actor.PlayerSession);
+
+            if(user is { } userId && component.Enabled)
+                _popup.PopupEntity(Loc.GetString("gas-analyzer-shutoff"), userId, Filter.Entities(userId));
+            component.Enabled = false;
+            SetAppearance(component);
+            Dirty(component);
+            _activeAnalyzers.Remove(uid);
+        }
+
+        /// <summary>
+        /// Disables the analyzer when the user closes the UI
+        /// </summary>
+        private void OnDisabledMessage(EntityUid uid, GasAnalyzerComponent component, GasAnalyzerDisableMessage message)
+        {
+            if (message.Session.AttachedEntity is not {Valid: true} user)
+                return;
+            // The UI should've already been closed but no harm in making sure?
+            DisableAnalyzer(uid, component, user);
+        }
+
+        private void OpenUserInterface(EntityUid user, GasAnalyzerComponent component)
+        {
+            if (!TryComp<ActorComponent>(user, out var actor))
+                return;
+
+            _userInterface.TryOpen(component.Owner, GasAnalyzerUiKey.Key, actor.PlayerSession);
+        }
+
+        /// <summary>
+        /// Fetches fresh data for the analyzer. Should only be called by Update or when the user requests an update via refresh button
+        /// </summary>
+        private bool UpdateAnalyzer(EntityUid uid, GasAnalyzerComponent? component = null)
+        {
+            if (!Resolve(uid, ref component))
+                return false;
+
+            // check if the user has walked away from what they scanned
+            var userPos = EntityManager.GetComponent<TransformComponent>(component.User).Coordinates;
```

Comment:
```suggestion
            var userPos = Transform(component.User).Coordinates;
```

etc for other uses of this--you can also use `Comp<TransformComponent>`. these are just entity-specific shorthands